test/libyul/KnowledgeBaseTest.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+    This file is part of solidity.
+
+    solidity is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    solidity is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Unit tests for KnowledgeBase
+ */
+
+#include <test/Common.h>
+
+#include <test/libyul/Common.h>
+
+#include <libyul/Object.h>
+#include <libyul/optimiser/KnowledgeBase.h>
+#include <libyul/optimiser/SSAValueTracker.h>
+#include <libyul/optimiser/DataFlowAnalyzer.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/CommonSubexpressionEliminator.h>
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <liblangutil/ErrorReporter.h>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace std;
+using namespace solidity::langutil;
+
+namespace solidity::yul::test
+{
+
+class KnowledgeBaseTest
+{
+protected:
+	KnowledgeBase constructKnowledgeBase(string const& _source)
+	{
+		ErrorList errorList;
+		shared_ptr<AsmAnalysisInfo> analysisInfo;
+		std::tie(m_object, analysisInfo) = yul::test::parse(_source, m_dialect, errorList);
+		BOOST_REQUIRE(m_object && errorList.empty() && m_object->code);
+
+		NameDispenser dispenser(m_dialect, *m_object->code);
+		std::set<YulString> reserved;
+		OptimiserStepContext context{m_dialect, dispenser, reserved, 0};
+		CommonSubexpressionEliminator::run(context, *m_object->code);
+
+		m_ssaValues(*m_object->code);
+		for (auto const& [name, expression]: m_ssaValues.values())
+			m_values[name].value = expression;
+
+		return KnowledgeBase(m_dialect, m_values);
+	}
+
+	EVMDialect m_dialect{EVMVersion{}, true};
+	shared_ptr<Object> m_object;
+	SSAValueTracker m_ssaValues;
+	map<YulString, AssignedValue> m_values;
+};
+
+BOOST_FIXTURE_TEST_SUITE(KnowledgeBase, KnowledgeBaseTest)
+
+BOOST_AUTO_TEST_CASE(basic)
+{
+//	yul::KnowledgeBase kb = constructKnowledgeBase(R"({
+//		let a := calldataload(0)
+//		let b := calldataload(0)
+//		let zero := 0
+//		let c := add(b, 0)
+//		let d := mul(b, 0)
+//		let e := sub(a, b)
+//	})");
+
+//	BOOST_CHECK(!kb.knownToBeDifferent("a"_yulstring, "b"_yulstring));
+//	// This only works if the variable names are the same.
+//	// It assumes that SSA+CSE+Simplifier actually replaces the variables.
```

Comment:
I think this can be improved by slightly modifying the pattern matching engine: https://github.com/ethereum/solidity/commit/8a94f173adf4ca154ef91b8bf7bad4b004efdc01

Reply:
The thing is that I actually don't want the pattern matching engine to be more complicated.